lib/quicly.c

```diff
@@ -4367,14 +4674,155 @@ static int handle_ping_frame(quicly_conn_t *conn, struct st_quicly_handle_payloa
+            conn->super.peer.stateless_reset.token = conn->super.peer.stateless_reset._buf;
+            conn->super.peer.cid_sequence = frame.sequence;
+            return 0;
+        }
+        /* replace with the picked one */
+        spare_cid->is_active = 0;
+        conn->super.peer.cid = spare_cid->cid;
+        memcpy(conn->super.peer.stateless_reset._buf, spare_cid->stateless_reset_token, QUICLY_STATELESS_RESET_TOKEN_LEN);
+        conn->super.peer.stateless_reset.token = conn->super.peer.stateless_reset._buf;
+        conn->super.peer.cid_sequence = spare_cid->sequence;
+    }
+
+    /* store new CID as a spare */
+    for (int i = 0; i < QUICLY_LOCAL_ACTIVE_CONNECTION_ID_LIMIT - 1; i++) {
+        struct st_quicly_spare_cid_t *spare_cid = &conn->super.peer.spare_cids[i];
+        if (spare_cid->is_active) {
+            /* If an endpoint receives a NEW_CONNECTION_ID frame that repeats a previously issued connection ID with
+             * a different Stateless Reset Token or a different sequence number, or if a sequence number is used for
+             * different connection IDs, the endpoint MAY treat that receipt as a connection error of type PROTOCOL_VIOLATION.
+             * (19.15)
+             */
+            if (cmp_cid(&spare_cid->cid, frame.cid) == 0) {
+                if (spare_cid->sequence == frame.sequence &&
+                    memcmp(spare_cid->stateless_reset_token, frame.stateless_reset_token, QUICLY_STATELESS_RESET_TOKEN_LEN) == 0) {
+                    /* likely a duplicate due to retransmission */
+                    return 0;
+                } else {
+                    /* received a frame that carries conflicting information */
+                    return QUICLY_TRANSPORT_ERROR_PROTOCOL_VIOLATION;
+                }
+            }
+            if (spare_cid->sequence == frame.sequence && cmp_cid(&spare_cid->cid, frame.cid) != 0)
+                return QUICLY_TRANSPORT_ERROR_PROTOCOL_VIOLATION;
+
+            continue;
+        }
+        spare_cid->sequence = frame.sequence;
+        set_cid(&spare_cid->cid, frame.cid);
+        memcpy(spare_cid->stateless_reset_token, frame.stateless_reset_token, QUICLY_STATELESS_RESET_TOKEN_LEN);
+        spare_cid->is_active = 1;
+        return 0;
+    }
+
+    /* Could not store the new CID -- peer sent NEW_CONNECTION_ID while we have already reached active_connection_id_limit */
+    return QUICLY_TRANSPORT_ERROR_CONNECTION_ID_LIMIT;
 }
 
 static int handle_retire_connection_id_frame(quicly_conn_t *conn, struct st_quicly_handle_payload_state_t *state)
 {
+    int ret;
+    quicly_retire_connection_id_frame_t frame;
+
+    if ((ret = quicly_decode_retire_connection_id_frame(&state->src, state->end, &frame)) != 0)
+        return ret;
+
+    /* TODO: return PROTOCOL_VIOLATION if sequence is associated with a DCID of this packet */
+
+    /* TODO: implement retirement logic */
```

Comment:
I now think we need to implement this before doing the next round of review. As stated above, this is inherently tied to how we handle the delivery of NCID frames.